ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -0,0 +1,441 @@
+ -                   |
+ -                   | ByteStrings
+ -                   |
+ -                   V
+ -           +-------+-------+
+ -           | Bearer.read() | MuxBearer implementation (Socket, Pipes, etc.)
+ -           +---------------+
+ -                   |
+ -                   | MuxSDUs
+ -                   |
+ -                   V
+ -           +-------+-------+
+ -           |     demux     | For a given MuxBearer there is a single demux thread reading from
+ -           +-------+-------+ the underlying berarer.
+ -                   |
+ -             +---+-+--+--+
+ -            /    |    |   \
+ -            v    v    v    v
+ -          |  | |  | |  | |  | There is a bounded queue for each mode (responder/initiator) per
+ -          |ci| |  | |bi| |br| miniprotocol.
+ -          |ci| |cr| |bi| |br|
+ -          +--+ +--+ +--+ +--+
+ -           |              |
+ -           | CBOR data    |
+ -           V              |
+ -       +---+-------+ Every ingress queueu has a dedicated thread which will read
+-        | muxDuplex | CBOR encoded data from its queue.
+-        | Initiator |      |
+-        | ChainSync |      |
+ -       +-----------+      |
+ -                          V
+ -                     +----+------+
+ -                     | muxDuplex |
+ -                     | Responder |
+ -                     | BlockFetch|
+ -                     +-----------+
+ -
+ -}
+
+-- | demux runs as a single separate thread and reads complete 'MuxSDU's from the underlying
+-- 'MuxBearer' and forwards it to the matching ingress queueu.
+demux :: forall m. (MuxBearer m, MonadSTM m, MonadSay m) => PerMuxSharedState m -> m ()
+demux pmss = forever $ do
+    (sdu, _) <- Ouroboros.Network.Mux.read (bearerHandle pmss)
+    --say $ printf "demuxing sdu on mid %s mode %s" (show $ msId sdu) (show $ msMode sdu)
+    -- Notice the mode reversal, ModeResponder is delivered to ModeInitiator and vice versa.
+    atomically $ writeTBQueue (ingressQueue (dispatchTable pmss) (msId sdu) (negMiniProtocolMode $ msMode sdu)) (msBlob sdu)
+
+-- | Return the ingress queueu for a given 'MiniProtocolId' and 'MiniProtocolMode'.
+ingressQueue :: (MuxBearer m) => MiniProtocolDispatch m -> MiniProtocolId -> MiniProtocolMode -> TBQueue m BL.ByteString
+ingressQueue (MiniProtocolDispatch tbl) dis mode =
+    case M.lookup (dis, mode) tbl of
+         Nothing -> error $ printf "Missing MiniProtocol %s mode %s in dispatch table"
+                                   (show dis) (show mode) -- XXX
+         Just q  -> q
+
+-- | muxJobs constructs a list of jobs which needs to be started in separate threads by
+-- the specific 'MuxBearer' instance.
+-- TODO: replace MonadSay with iohk-monitoring-framework.
+muxJobs :: (MuxBearer m, MonadSTM m, MonadSay m) =>
+    MiniProtocolDescriptions m ->
+    MuxBearerHandle m ->
+    m [m ()]
+muxJobs (MiniProtocolDescriptions udesc) bearer = do
+    tbl <- setupTbl
+    tq <- atomically $ newTBQueue 100
```

Comment:
If I am not mistaken this list will be never longer than the number of miniprotocols.  The reason is that they will all block on writing to a wanton first before placing it on the list, @karknu am I possibly right?

Reply:
That is correct.

Reply:
Maybe add a comment on that, it's a useful operational information.